File: backend/workflow/pool.go
package workflow

import (
	"context"
	"strings"

	"github.com/enough/enough/backend/core"
)

type poolResult struct {
	job    queuedJob
	result AgentResult
}

func (r *Runtime) runPool(ctx context.Context, phase string, jobs []queuedJob) ([]AgentResult, error) {
	if len(jobs) == 0 {
		return nil, nil
	}
	r.registerJobs(phase, jobs)
	pending := append([]queuedJob(nil), jobs...)
	results := make([]AgentResult, 0, len(jobs))
	resultCh := make(chan poolResult, len(jobs)+r.maxConcurrency)
	running := 0

	for len(pending) > 0 || running > 0 {
		if err := ctx.Err(); err != nil {
			r.cancelActiveAgents()
			drainPoolResults(resultCh, running)
			return results, err
		}

		r.mu.RLock()
		paused := r.paused
		concurrency := r.maxConcurrency
		r.mu.RUnlock()

		for !paused && running < concurrency && len(pending) > 0 {
			job := pending[0]
			pending = pending[1:]
			if completed, ok := r.completedResult(job.Key); ok {
				results = append(results, completed)
				r.markCompletedFromCheckpoint(job, completed)
				continue
			}
			running++
			go func(job queuedJob) {
				resultCh <- poolResult{job: job, result: r.executeJob(ctx, phase, job.Key, job.Options)}
			}(job)
		}

		if running == 0 {
			if len(pending) == 0 {
				break
			}
			select {
			case <-ctx.Done():
				r.cancelActiveAgents()
				return results, ctx.Err()
			case <-r.wake:
				continue
			}
		}

		select {
		case <-ctx.Done():
			r.cancelActiveAgents()
			drainPoolResults(resultCh, running)
			return results, ctx.Err()
		case <-r.wake:
			continue
		case item := <-resultCh:
			running--
			if r.takeRestart(item.job.Key) {
				r.resetQueued(item.job)
				pending = append([]queuedJob{item.job}, pending...)
				continue
			}
			if isQuotaError(item.result.Error) {
				r.pauseForQuota()
				r.cancelActiveAgents()
				drainPoolResults(resultCh, running)
				return results, ErrQuotaPaused
			}
			results = append(results, item.result)
		}
	}
	return results, nil
}

func drainPoolResults(resultCh <-chan poolResult, count int) {
	for count > 0 {
		<-resultCh
		count--
	}
}

func (r *Runtime) registerJobs(phase string, jobs []queuedJob) {
	r.mu.Lock()
	r.ensurePhaseLocked(phase)
	for _, job := range jobs {
		if _, exists := r.snapshot.Agents[job.Key]; exists {
			continue
		}
		r.snapshot.Agents[job.Key] = AgentSnapshot{
			Key: job.Key, Phase: phase, Role: defaultRole(job.Options.Role),
			Status: "queued", Prompt: job.Options.Prompt,
		}
	}
	r.recountLocked()
	r.mu.Unlock()
	r.persist()
	r.emitRun(core.EventWorkflowPhase, phase)
}

func (r *Runtime) completedResult(key string) (AgentResult, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	result, ok := r.completed[key]
	return result, ok
}

func (r *Runtime) markCompletedFromCheckpoint(job queuedJob, result AgentResult) {
	r.mu.Lock()
	s := r.snapshot.Agents[job.Key]
	s.Status = "done"
	if !result.OK {
		s.Status = "failed"
	}
	s.Result, s.JSON, s.Error = result.Text, result.JSON, result.Error
	s.Tokens, s.Turns = result.TokensUsed, result.TurnCount
	r.snapshot.Agents[job.Key] = s
	r.recountLocked()
	r.mu.Unlock()
}

func (r *Runtime) takeRestart(key string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	restart := r.restartRequested[key]
	delete(r.restartRequested, key)
	return restart
}

func (r *Runtime) resetQueued(job queuedJob) {
	r.mu.Lock()
	s := r.snapshot.Agents[job.Key]
	s.Status = "queued"
	s.Error = ""
	s.Result = ""
	s.JSON = nil
	r.snapshot.Agents[job.Key] = s
	delete(r.completed, job.Key)
	r.recountLocked()
	r.mu.Unlock()
	r.persist()
}

func (r *Runtime) pauseForQuota() {
	r.mu.Lock()
	r.paused = true
	r.pauseReason = "provider quota exceeded"
	r.snapshot.Status = "paused"
	r.snapshot.Message = "provider quota exceeded"
	if r.state != nil {
		r.state.Status = "paused"
		r.state.PauseReason = "provider quota exceeded"
	}
	r.mu.Unlock()
	r.persist()
}

func (r *Runtime) cancelActiveAgents() {
	r.mu.RLock()
	var cancels []context.CancelFunc
	for _, control := range r.active {
		cancels = append(cancels, control.cancel)
	}
	r.mu.RUnlock()
	for _, cancel := range cancels {
		cancel()
	}
}

func (r *Runtime) ensurePhaseLocked(name string) {
	for _, phase := range r.snapshot.Phases {
		if phase.Name == name {
			return
		}
	}
	r.snapshot.Phases = append(r.snapshot.Phases, PhaseSnapshot{Name: name})
}

func (r *Runtime) recountLocked() {
	r.snapshot.Queued, r.snapshot.Running, r.snapshot.Done, r.snapshot.Failed, r.snapshot.Tokens = 0, 0, 0, 0, 0
	phases := map[string]*PhaseSnapshot{}
	for i := range r.snapshot.Phases {
		p := &r.snapshot.Phases[i]
		p.Total, p.Queued, p.Running, p.Done, p.Failed, p.Tokens = 0, 0, 0, 0, 0, 0
		phases[p.Name] = p
	}
	for _, agent := range r.snapshot.Agents {
		phase := phases[agent.Phase]
		if phase == nil {
			continue
		}
		phase.Total++
		phase.Tokens += agent.Tokens
		r.snapshot.Tokens += agent.Tokens
		switch agent.Status {
		case "queued":
			phase.Queued++
			r.snapshot.Queued++
		case "running":
			phase.Running++
			r.snapshot.Running++
		case "done":
			phase.Done++
			r.snapshot.Done++
		case "failed", "stopped":
			phase.Failed++
			r.snapshot.Failed++
		}
	}
}

func isQuotaError(text string) bool {
	if text == "" {
		return false
	}
	for _, needle := range []string{"rate limit", "rate_limit", "quota", "usage limit", "too many requests", "status 429", "http 429"} {
		if containsFold(text, needle) {
			return true
		}
	}
	return false
}

func containsFold(s, needle string) bool {
	return strings.Contains(strings.ToLower(s), strings.ToLower(needle))
}

File: backend/workflow/pool.ts
// PORT: backend/workflow/pool.go

import { Effect } from "effect";
import { type AgentResult, type PhaseSnapshot, type AgentSnapshot, cloneJSON } from "./types";
import { type queuedJob } from "./pipeline";
import { Runtime } from "./runtime";
import { event_workflow_phase } from "../core/events";
import { defaultRole } from "./spawn";

export class QuotaError extends Error {
  constructor() {
    super("provider quota exceeded");
    this.name = "QuotaError";
  }
}

export function runPool(
  this: Runtime,
  ctx: AbortSignal,
  phase: string,
  jobs: queuedJob[]
): Effect.Effect<AgentResult[], Error> {
  if (jobs.length === 0) {
    return Effect.succeed([]);
  }

  return Effect.async<AgentResult[], Error>((resume) => {
    this.registerJobs(phase, jobs);

    let pending = [...jobs];
    const results: AgentResult[] = [];
    let running = 0;
    const runningJobs = new Map<string, Promise<void>>();

    let finishedResolver: (() => void) | null = null;
    let wakeResolver: (() => void) | null = null;

    const onAbort = () => {
      this.cancelActiveAgents();
      resume(Effect.fail(new Error("interrupted")));
    };

    if (ctx.aborted) {
      onAbort();
      return;
    }
    ctx.addEventListener("abort", onAbort);

    const wakeCallback = () => {
      if (wakeResolver) {
        wakeResolver();
      }
    };
    this.wakeCallbacks.add(wakeCallback);

    const cleanup = () => {
      ctx.removeEventListener("abort", onAbort);
      this.wakeCallbacks.delete(wakeCallback);
    };

    const loop = async () => {
      try {
        while (pending.length > 0 || running > 0) {
          if (ctx.aborted) {
            this.cancelActiveAgents();
            throw new Error("interrupted");
          }

          const paused = this.paused;
          const concurrency = this.maxConcurrency;

          // Start new jobs if we have capacity and are not paused
          while (!paused && running < concurrency && pending.length > 0) {
            const job = pending.shift()!;
            const completed = this.completedResult(job.Key);
            if (completed) {
              results.push(completed);
              this.markCompletedFromCheckpoint(job, completed);
              continue;
            }

            running++;
            const jobPromise = (async () => {
              try {
                const res = await this.executeJob(ctx, phase, job.Key, job.Options);
                running--;
                runningJobs.delete(job.Key);

                if (this.takeRestart(job.Key)) {
                  this.resetQueued(job);
                  pending.unshift(job); // Add to the front
                } else if (isQuotaError(res.error || "")) {
                  this.pauseForQuota();
                  this.cancelActiveAgents();
                  throw new QuotaError();
                } else {
                  results.push(res);
                }
              } catch (e) {
                throw e;
              } finally {
                if (finishedResolver) {
                  finishedResolver();
                }
              }
            })();

            runningJobs.set(job.Key, jobPromise);
          }

          if (running === 0) {
            if (pending.length === 0) {
              break;
            }
            // If we're paused and no jobs are running but there are pending jobs,
            // we must wait for a wake signal.
            await new Promise<void>((resolve) => {
              wakeResolver = resolve;
            });
            wakeResolver = null;
            continue;
          }

          // Wait until either a job finishes or a wake signal occurs
          await Promise.race([
            new Promise<void>((resolve) => {
              finishedResolver = resolve;
            }),
            new Promise<void>((resolve) => {
              wakeResolver = resolve;
            })
          ]);
          finishedResolver = null;
          wakeResolver = null;
        }

        cleanup();
        resume(Effect.succeed(results));
      } catch (err: any) {
        cleanup();
        if (err instanceof QuotaError) {
          resume(Effect.fail(new Error("workflow paused: provider quota exceeded")));
        } else {
          resume(Effect.fail(err));
        }
      }
    };

    loop();
  });
}

export function registerJobs(this: Runtime, phase: string, jobs: queuedJob[]): void {
  this.ensurePhaseLocked(phase);
  for (const job of jobs) {
    if (this.snapshot.agents[job.Key]) {
      continue;
    }
    this.snapshot.agents[job.Key] = {
      key: job.Key,
      phase: phase,
      role: defaultRole(job.Options.role ?? ""),
      status: "queued",
      prompt: job.Options.prompt,
    };
  }
  this.recountLocked();
  this.persist();
  this.emitRun(event_workflow_phase, phase);
}

export function completedResult(this: Runtime, key: string): AgentResult | undefined {
  return this.completed[key];
}

export function markCompletedFromCheckpoint(this: Runtime, job: queuedJob, result: AgentResult): void {
  const s = this.snapshot.agents[job.Key] || {
    key: "",
    phase: "",
    role: "",
    status: "",
    prompt: "",
  };
  s.status = "done";
  if (!result.ok) {
    s.status = "failed";
  }
  s.result = result.text;
  s.json = result.json;
  s.error = result.error ?? "";
  s.tokens = result.tokensUsed;
  s.turns = result.turnCount;
  this.snapshot.agents[job.Key] = s;
  this.recountLocked();
}

export function takeRestart(this: Runtime, key: string): boolean {
  const restart = !!this.restartRequested[key];
  delete this.restartRequested[key];
  return restart;
}

export function resetQueued(this: Runtime, job: queuedJob): void {
  const s = this.snapshot.agents[job.Key] || {
    key: "",
    phase: "",
    role: "",
    status: "",
    prompt: "",
  };
  s.status = "queued";
  s.error = "";
  s.result = "";
  s.json = undefined;
  this.snapshot.agents[job.Key] = s;
  delete this.completed[job.Key];
  this.recountLocked();
  this.persist();
}

export function pauseForQuota(this: Runtime): void {
  this.paused = true;
  this.pauseReason = "provider quota exceeded";
  this.snapshot.status = "paused";
  this.snapshot.message = "provider quota exceeded";
  if (this.state) {
    this.state.status = "paused";
    this.state.pauseReason = "provider quota exceeded";
  }
  this.persist();
}

export function cancelActiveAgents(this: Runtime): void {
  const cancels = Object.values(this.active).map((c) => c.cancel);
  for (const cancel of cancels) {
    try {
      cancel();
    } catch {}
  }
}

export function ensurePhaseLocked(this: Runtime, name: string): void {
  for (const phase of this.snapshot.phases) {
    if (phase.name === name) {
      return;
    }
  }
  this.snapshot.phases.push({
    name,
    total: 0,
    queued: 0,
    running: 0,
    done: 0,
    failed: 0,
    tokens: 0,
  });
}

export function recountLocked(this: Runtime): void {
  this.snapshot.queued = 0;
  this.snapshot.running = 0;
  this.snapshot.done = 0;
  this.snapshot.failed = 0;
  this.snapshot.tokens = 0;

  const phases: Record<string, PhaseSnapshot> = {};
  for (const p of this.snapshot.phases) {
    p.total = 0;
    p.queued = 0;
    p.running = 0;
    p.done = 0;
    p.failed = 0;
    p.tokens = 0;
    phases[p.name] = p;
  }

  for (const agent of Object.values(this.snapshot.agents)) {
    const phase = phases[agent.phase];
    if (!phase) {
      continue;
    }
    phase.total++;
    phase.tokens += agent.tokens ?? 0;
    this.snapshot.tokens += agent.tokens ?? 0;
    switch (agent.status) {
      case "queued":
        phase.queued++;
        this.snapshot.queued++;
        break;
      case "running":
        phase.running++;
        this.snapshot.running++;
        break;
      case "done":
        phase.done++;
        this.snapshot.done++;
        break;
      case "failed":
      case "stopped":
        phase.failed++;
        this.snapshot.failed++;
        break;
    }
  }
}

export function isQuotaError(text: string): boolean {
  if (text === "") {
    return false;
  }
  const needles = ["rate limit", "rate_limit", "quota", "usage limit", "too many requests", "status 429", "http 429"];
  const lowerText = text.toLowerCase();
  for (const needle of needles) {
    if (lowerText.includes(needle)) {
      return true;
    }
  }
  return false;
}

// Attach prototype methods
const proto = Runtime.prototype as any;
proto.runPool = runPool;
proto.registerJobs = registerJobs;
proto.completedResult = completedResult;
proto.markCompletedFromCheckpoint = markCompletedFromCheckpoint;
proto.takeRestart = takeRestart;
proto.resetQueued = resetQueued;
proto.pauseForQuota = pauseForQuota;
proto.cancelActiveAgents = cancelActiveAgents;
proto.ensurePhaseLocked = ensurePhaseLocked;
proto.recountLocked = recountLocked;

/*
PORT STATUS
source path: backend/workflow/pool.go
source lines: 241
draft lines: 290
confidence: high
status: phase_b_compile
*/
